### PR TITLE
[OpenACC] Implement beginning parts of the 'parallel' Sema impl

### PR DESCRIPTION
**Pull request by erichkeane**
This patch Implements AST node creation and appertainment enforcement for 'parallel', as well as changes the 'not implemented' messages to be more specific.  It does not deal with clauses/clause legality, nor a few of the other rules from the standard, but this gets us most of the way for a framework for future construct implementation.

### CODE REVIEW

File: clang/lib/Sema/SemaOpenACC.cpp [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+//===----------------------------------------------------------------------===//
+
+#include "clang/Basic/DiagnosticSema.h"
+#include "clang/Basic/OpenACCKinds.h"
+#include "clang/Sema/Sema.h"
+
+using namespace clang;
+
+namespace {
+bool DiagnoseConstructAppertainment(Sema &S, OpenACCDirectiveKind K,
```
Comment (alexey-bataev) [inline clang/lib/Sema/SemaOpenACC.cpp:21]: diagnoseConstructAppertainment

File: clang/lib/Sema/SemaOpenACC.cpp [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+                                    SourceLocation StartLoc, bool IsStmt) {
+  switch (K) {
+  default:
+  case OpenACCDirectiveKind::Invalid:
+    // Nothing to do here, both invalid and unimplemented don't really need to
+    // do anything.
+    break;
```
Comment (alexey-bataev) [inline clang/lib/Sema/SemaOpenACC.cpp:28]: Should it be llvm_unreachable instead?

File: clang/lib/Parse/ParseOpenACC.cpp [hunk 3]
```diff
@@ -745,9 +745,14 @@ bool Parser::ParseOpenACCClause(OpenACCDirectiveKind DirKind) {
...
   // Consume the clause name.
-  ConsumeToken();
+  SourceLocation ClauseLoc = ConsumeToken();
+
+  bool ParamsResult = ParseOpenACCClauseParams(DirKind, Kind);
 
-  return ParseOpenACCClauseParams(DirKind, Kind);
+  // TODO OpenACC: this whole function should return a 'clause' type optional
+  // instead of bool, so we likely want to return the clause here.
+  getActions().ActOnOpenACCClause(Kind, ClauseLoc);
```
Comment (alexey-bataev) [inline clang/lib/Parse/ParseOpenACC.cpp:754]: This one does not nothing right now, just emit error message. Can you split it and implement in a separate patch?

File: clang/lib/AST/ASTContext.cpp [hunk 4]
```diff
@@ -13662,3 +13663,8 @@ StringRef ASTContext::getCUIDHash() const {
   CUIDHash = llvm::utohexstr(llvm::MD5Hash(LangOpts.CUID), /*LowerCase=*/true);
   return CUIDHash;
 }
+
+void ASTContext::setOpenACCStructuredBlock(OpenACCComputeConstruct *C,
+                                           Stmt *S) {
+  C->setStructuredBlock(S);
+}
```
Comment (alexey-bataev) [inline clang/lib/AST/ASTContext.cpp:13670]: I don't like this approach, better to set structured block upon OpenACCComputeConstruct construction, if possible.

File: clang/lib/Sema/SemaOpenACC.cpp [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+  // declaration. At the moment, only handle appertainment.
+  DiagnoseConstructAppertainment(*this, K, StartLoc, /*IsStmt=*/false);
+}
+
+void Sema::ActOnEndOpenACCDeclDirective() {
+  // TODO OpenACC: Should diagnose anything having to do with the associated
+  // statement, or any clause diagnostics that can only be done at the 'end' of
+  // the directive.  We should also close any 'block' marking now that the decl
+  // parsing is complete.
+}
```
Comment (alexey-bataev) [inline clang/lib/Sema/SemaOpenACC.cpp:83]: Can you try to introduce all these empty functions in a separate NFC patch instead?

File: clang/lib/Parse/ParseOpenACC.cpp (hunk 3 above)
Comment (erichkeane) [reply clang/lib/Parse/ParseOpenACC.cpp:754]: Sure, I can do that.

File: clang/lib/Sema/SemaOpenACC.cpp (hunk 5 above)
Comment (erichkeane) [reply clang/lib/Sema/SemaOpenACC.cpp:83]: In this case, it ends up being a 'start' function without an 'end' if we do that.  That said, your suggestion above was to effectively no longer do the 'start' and 'end' by delaying construction until after parsing, so I might be able to combine the 'ActOnStart' and 'ActOnEnd' into 1 anyway.

Comment (erichkeane) [reply clang/lib/Sema/SemaOpenACC.cpp:83]: Ah, I think I see what you mean.  I'll submit a patch that adds just the sema functions and calls them properly, then can add the "change the warnings" as a separate patch, then the "implement parallel" as a patch.

Comment (alexey-bataev) [reply clang/lib/Sema/SemaOpenACC.cpp:83]: Yep, just separate structural and functional changes.

File: clang/lib/Sema/SemaOpenACC.cpp (hunk 2 above)
Comment (alexey-bataev) [reply clang/lib/Sema/SemaOpenACC.cpp:28]: Should it be llvm_unreachable?

Comment (erichkeane) [reply clang/lib/Sema/SemaOpenACC.cpp:28]: No, that would just end up crashing, we're calling into the 'act on start' even when invalid, which should allow us to handle 'state' of bad pragmas better, so 'invalid' has to get through here without causing problems.  The intent is that we wont create an AST node here, but still allow 'checking' of everything else.

**Review by alexey-bataev — APPROVED** [review]
LG
